SSA/Projects/RISCV64/Base.lean

```diff
@@ -0,0 +1,390 @@
+-/
+inductive Op
+| li : (val : Int) → Op
+|  lui (imm : BitVec 20)
+|  auipc (imm : BitVec 20)
+|  addi (imm : BitVec 12)
+|  andi (imm : BitVec 12)
+|  ori (imm : BitVec 12)
+|  xori (imm : BitVec 12)
+|  addiw (imm : BitVec 12)
+|  add
+|  slli (shamt : BitVec 6)
+|  sub
+|  and
+|  or
+|  xor
+|  sll
+|  srl
+|  sra
+|  addw
+|  subw
+|  sllw
+|  srlw
+|  sraw
+-- fence
+|  slti (imm : BitVec 12)
+|  sltiu (imm : BitVec 12)
+|  srli (shamt : BitVec 6)
+|  srai (shamt : BitVec 6)
+|  slliw (shamt : BitVec 5)
+|  srliw (shamt : BitVec 5)
+|  sraiw (shamt : BitVec 5)
+|  slt
+|  sltu
+-- RISC-V `M` extension instructions (multiply& divide)
+| mul    -- performs signed multiplication on 64 x 64 bits and returns the lower 64 bits of the result .
+| mulu   -- performs unsigned multiplication on 64 x 64 bits and returns the lower 64 bits of the result .
+| mulw
+| mulh   -- performs signed multiplication on 64 x 64 bits and returns the upper 64 bits of the result.
+| mulhu  -- performs unsigned multiplication on 64 x 64 bits and returns the upper 64 bits of the result.
+| mulhsu -- performs multiplication on (rs1 signed) x (rs2 unsigned) and returns the upper 64 bits of the result.
+| divw
+| divwu
+| div    -- signed division
+| divu   -- unsigned division
+| remw
+| rem    --sign of result is according to sign of dividend
+| remwu
+| remu
+-- RISC-V `Zba` extension
+|add.uw
+|sh1add.uw
+|sh2add.uw
+|sh3add.uw
+|sh1add
+|sh2add
+|sh3add
+-- slli.uw missing
+-- part of the RISC-V `Zbb` & `Zbkb` extension
+| rol
+| ror
+| rolw
+| rorw
+| sext.b
+| sext.h
+| zext.h
+/-
+in the future:
+|pack
+|packh
+-/
+-- part of the RISC-V `Zbs` extension
+| bclr
+| bclri (shamt : BitVec 6)
+| bext
+| bexti (shamt : BitVec 6)
+| binv
+| binvi (shamt : BitVec 6)
+| bset
+| bseti (shamt : BitVec 6)
+/- RISC-V `Zicond` conditional operations extension  -/
+| czero.eqz
+| czero.nez
+deriving DecidableEq, Repr
+
+/--
+## Dialect type definitions
+Defining a type system for the `RISCV64` dialect. `bv` represents bit vector.
+-/
+inductive Ty
+  | bv : Ty
+  deriving DecidableEq, Repr, Inhabited
+
+instance : ToString (Ty) where
+  toString t := repr t |>.pretty
+
+/-!
+Connecting the `bv` type to its underlying Lean type `BitVec 64`. By providing a `TyDenote` instance,
+we define how the `RISCV64` types transalte into actual Lean types.
+-/
+instance : TyDenote Ty where
+toType := fun
+| Ty.bv => BitVec 64
+
+instance (ty : Ty) : Inhabited (TyDenote.toType ty) where
+  default := match ty with
+    | .bv  => 0#64
+
+/--
+## Dialect operation definitions
+
+Specifing the signature of each `RISCV64` operation. `Sig` refers to the input types
+for each operation as a list of types.
+
+We mark it as `simp` and `reducible` such that the type checker and elaborator
+always fully unfold a `sig` to its underlying defintion and when the simplifier
+encounters a `sig` it can replace it by its definition.
```

Comment:
```suggestion
always fully unfold a `sig` to its underlying definition and when the simplifier
encounters a `sig` it can replace it by its definition.
```